### DISCUSSION TITLE
useSession/SessionProvider in Stories

### DISCUSSION THREAD

**Discussion by aalega2** [nextauthjs/next-auth#4979] [Help]
Hi all,

I'm using Storybook for UI components and some of my components call `useSession()`. I tried setting a decorator that wrapped around the component and globally to no avail.

Any clue on how to mock this out? Thanks!

**Comment by bizk-sato** [comment]
@aalega2 
Is there any solution for this?

**Comment by fivaz1** [comment]
Still waiting for a solution

**Comment by TomFreudenberg** [comment]
:-) Yes there is something "coming"

The applied patch https://github.com/nextauthjs/next-auth/pull/5438 is exactly what you need for that.

This is how my storybook provider looks like with the patch applied:

```js
import PropTypes from 'prop-types';

import { SessionContext } from 'next-auth/react';
import { useMemo } from 'react';
import { jsx as _jsx } from "react/jsx-runtime";

const SessionContextProvider = ({ children, session }) => {
  const value = useMemo(() => {
    return session ? session : { data: undefined, status: "unauthenticated" };
  }, [session]);

  return (_jsx(SessionContext.Provider, { value: value, children: children }));
}

SessionContextProvider.propTypes = {
  children: PropTypes.node.isRequired
};

export default SessionContextProvider;
```

and to use it in storybook I made this setup in `preview.js`:

```js
// apply the providers to stories
const withAppProviders = (Story, context) => {
  // define properties for components and appStates
  const session = previewAuthentications[context.globals.appAuthenticated]?.session;
  const locale = context.globals.appLocale;
  const theme = context.globals.appTheme;
  const appStates = {
    theme: {
      active: theme,
      isLightMode: theme === THEME.LIGHT_IDENTIFIER,
      isDarkMode: theme === THEME.DARK_IDENTIFIER
    },
    i18n: {
      locale: locale
    }
  };

  return (
    <SessionContextProvider session={session}>
      <SbAppContextProvider storedAppStates={appStates}>
        <ApplyLocale locale={locale} />
        <LocaleContextProvider>
          <ApplyTheme theme={theme} />
          <ThemeContextProvider injectFirst>
            <Story {...context} />
          </ThemeContextProvider>
        </LocaleContextProvider>
      </SbAppContextProvider>
    </SessionContextProvider>
  );
};

// push to decorators
export const decorators = [withAppProviders, ...RouterDecorators];
```

checkout the video in https://github.com/nextauthjs/next-auth/discussions/5437

that should be what you are looking for?

**Reply by TomFreudenberg** [reply]
@balazsorban44 do you already have some idea/time when releasing a new version where the patch is applied?

**Reply by fivaz1** [reply]
have you published your fork of next auth on npm or yarn ?

**Reply by TomFreudenberg** [reply]
Hi @fivaz, not separated deployed but you may use the next-auth github repo if won't wait.

From CONTRIBUTION.md

```bash

git clone git@github.com:nextauthjs/next-auth.git

cd next-auth

pnpm install

pnpm build

```

Now `cd` into your application and link the local next-auth

```bash

cd .../my-application

cd ./node_modules

rm -rf ./next-auth

ln -s /full-path/next-auth/packages/next-auth .

cd ..

```

_Annotation: If using pnpm for your application you may use `pnpm link` instead of linking from shell_

You should be enabled now to use SessionContext as above.

**Reply by fivaz1** [reply]
Thank you

**Reply by TomFreudenberg** [reply]
If interested in my globalTypes:

```js
// get additional settings for authentication
import previewAuthentications from './previewAuthentications';

// define global options for theme and locale switcher
export const globalTypes = {
  appAuthenticated: {
    name: 'appAuthenticated',
    description: 'Set authentication state',
    defaultValue: null,
    toolbar: {
      icon: 'user',
      items: Object.keys(previewAuthentications).map((u) => ({ value: u, title: previewAuthentications[u].title }))
    }
  }
};
```

and my `./.storybook/previewAuthentications.js` file.

you may define additional properties like roles etc. as well in here to the session record.

```js
// some standard viewports to check a page and story
// on different well-known devices

module.exports = {
  unknown: {
    title: 'session unknown',
    session: null
  },
  loading: {
    title: 'session loading',
    session: {
      data: null,
      status: 'loading'
    }
  },
  admin: {
    title: 'admin not authenticated',
    session: {
      data: {
        id: 1,
        login: 'admin',
        username: 'Administrator',
        email: 'admin@local'
      },
      status: 'unauthenticated'
    }
  },
  adminAuthed: {
    title: 'admin authenticated',
    session: {
      data: {
        id: 1,
        login: 'admin',
        username: 'Administrator',
        email: 'admin@local'
      },
      status: 'authenticated'
    }
  },
  user: {
    title: 'user not authenticated',
    session: {
      data: {
        id: 999,
        login: 'user',
        username: 'User',
        email: 'user@local'
      },
      status: 'unauthenticated'
    }
  },
  userAuthed: {
    title: 'user authenticated',
    session: {
      data: {
        id: 999,
        login: 'user',
        username: 'User',
        email: 'user@local'
      },
      status: 'authenticated'
    }
  }
};
```

**Reply by fivaz1** [reply]
thanks man I'm gonna test it

**Comment by TomFreudenberg** [comment]
Hi @balazsorban44 - maybe you or @aalega2 may change the discussion title into:

"How to use Next-Auth SessionContext and useSession in Storybook"

to be more precise and help "searching"

**Comment by TomFreudenberg** [comment]
Well :-) here is a component which does all in.

https://github.com/TomFreudenberg/next-auth-mock

If please the new release may be pushed - it would be very easy to use Next-Auth also in Storybook mocking.

Just install `npm install --save-dev @tomfreudenberg/next-auth-mock`

**Reply by TomFreudenberg** [reply]
The new release 4.12.3 is out and never get easier integrated into storybook.

<img width="514" alt="image" src="https://user-images.githubusercontent.com/410087/194185214-75dc9e75-75cf-425c-973f-583584d7fafb.png">

**Reply by TomFreudenberg** [reply]
Its now more easy than ever :-)

Just add this to your storybook addons list:

```js
module.exports = {
  addons: ['@tomfreudenberg/next-auth-mock/storybook']
}
```

Package: https://npmjs.com/package/@tomfreudenberg/next-auth-mock

**Reply by st4l1nR** [reply]
I tried this config but im still recibed the  "[next-auth]: `useSession` must be wrapped in a <SessionProvider />"  message

**Reply by TomFreudenberg** [reply]
Hi @st4l1nR - thanks for opening the issue in https://github.com/TomFreudenberg/next-auth-mock/issues/2

**Reply by dcarias0388** [reply]
> Probé esta configuración, pero todavía recibí el "[next-auth]: `useSession`debe estar envuelto en un "mensaje
  
Ya logro solucionar su error?...a mi me muestra el mismo error, y es porque en mi componente obtengo la session usando el useSession(). Igualmente me muestra [next-auth]: `useSession` must be wrapped in a <SessionProvider />

**Reply by TomFreudenberg** [reply]
Hi @dcarias0388 Daileny

would you please post an issue here: https://github.com/TomFreudenberg/next-auth-mock/issues

If possible, let me checkout your code

Cheers, 
Tom

**Reply by sneko** [reply]
@TomFreudenberg thanks for your plugin.

Any idea how to mock the `next-auth` methods? For example I have some components that use `signIn(...) / signOut(...)` and in my Storybook it triggers fatal redirections. I would like to make them doing nothing at all.

... it would be better if in the JS world people would think about real mocking from start (cf. `next-auth`, `next`, `prisma`, `trpc`). It's so painful to patch things :D and it's pretty unstable over time.

(for sure I still have the possibility to make "homemade proxies of export/import" with conditions behind, maybe that's what I should adopt ^^, like that it would use a logicless `signIn(...)` for example)

**Reply by TomFreudenberg** [reply]
@dcarias0388 @st4l1nR today a new release was posted - removing dependencies which were causing your issue. If you could update and check, it should work now without the error `useSession must be wrapped ...` Thanks for some feedback

**Comment by TomFreudenberg** [comment]
Hi Thomas @sneko

I have not yet thought about that issue. A simple one could be managed be setting a property like `testing: true` into your mockup-data and check that `flag` in your component. Maybe better than proxing etc. ?
